README.md
# rescue-api

* Documentation
  * [Official][0]
  * [Unofficial][1]

[0]: https://www.rescuetime.com/anapi/setup/documentation
[1]: https://man.sr.ht/~mjorgensen/rescuetime-api-docs/

This is a Python 3 implementation of the RescueTime API.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic
Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- `install_requires` key-value in `setup.py` to let pip know to install requests
  as this API wrapper depends on it!

### Changed


## [1.0.2]
### Changed
- Added an `__init__.py` to make `rescue_api/` an actual Python package. This
  resolves all import issues.


## [1.0.1]
### Changed
- Modified `setup.py` in an effort to make imports work.

## [1.0.0]
### Added
- Fully implemented RescueTime API (to the extent of their documentation!)
- Added package to PyPi - <https://pypi.org/project/rescue-api/>

## [0.9.0] - 2019-01-25
### Added
- Implemented support for the rest of the RescueTime API
  - Daily Summary API!
  - Alerts Feed API!
  - Highlights Feed API!
  - Highlights Post API!

[Unreleased]: https://git.sr.ht/~mjorgensen/rescue-api/
[1.0.2]: https://git.sr.ht/~mjorgensen/rescue-api/refs/1.0.2
[1.0.1]: https://git.sr.ht/~mjorgensen/rescue-api/refs/1.0.1
[1.0.0]: https://git.sr.ht/~mjorgensen/rescue-api/refs/1.0.0
[0.9.0]: https://git.sr.ht/~mjorgensen/rescue-api/refs/0.9.0

run_tests.py
import json
import rescuetime

rt = rescuetime.RescueAPI()

print("""
--------------------------------------------------------------------------------
Testing Analytic Data API
--------------------------------------------------------------------------------
""")
result = rt.get_analytic_data()
print(result['status_code'])
if result['format'] == 'json':
    payload = json.dumps(result['data'], indent=2)
if result['format'] == 'csv':
    payload = result['data']
print(payload)

print("""
--------------------------------------------------------------------------------
Testing Daily Summary Feed
--------------------------------------------------------------------------------
""")
result = rt.get_daily_summary_feed()
print(result['status_code'])
payload = json.dumps(result['json'], indent=2)
print(payload)

print("""
--------------------------------------------------------------------------------
Testing Alerts Feed
--------------------------------------------------------------------------------
""")
result = rt.list_alerts()
print(result['status_code'])
payload = json.dumps(result['json'], indent=2)
print(payload)

print("""
--------------------------------------------------------------------------------
Testing Alerts Status
--------------------------------------------------------------------------------
""")
result = rt.alert_status()
print(result['status_code'])
payload = json.dumps(result['json'], indent=2)
print(payload)

print("""
--------------------------------------------------------------------------------
Testing Highlights Feed
--------------------------------------------------------------------------------
""")
result = rt.read_highlights(end_date="2019-01-23")
print(result['status_code'])
payload = json.dumps(result['json'], indent=2)
print(payload)

print("""
--------------------------------------------------------------------------------
Testing Highlights POST
--------------------------------------------------------------------------------
""")
result = rt.post_highlight("Test")
payload = result['json']
if result['status_code'] == 200:
    print('Success!')
    print(payload)
elif result['status_code'] == 400:
    print('Fail!')
    print(payload)


rescuetime.py
import os
import datetime
import requests


class RescueAPI:

    def __init__(self, ):
        self.API_KEY = os.environ['RESCUETIME_API_KEY']
        self.payload = {"key": self.API_KEY}

    def endpoint(self, endpoint):
        base_url = "https://www.rescuetime.com/anapi"
        endpoint = base_url + endpoint
        return endpoint

    def get_analytic_data(self, format="json"):
        _endpoint = self.endpoint("/data")
        payload = self.payload
        payload['format'] = format
        r = requests.get(_endpoint, params=payload)
        if format.lower() == "json":
            return {'status_code': r.status_code, 'data': r.json()}
        if format.lower() == "csv":
            raise NotImplementedError

    def get_daily_summary_feed(self,):
        _endpoint = self.endpoint("/daily_summary_feed")
        payload = self.payload
        r = requests.get(_endpoint, params=payload)
        return {'status_code': r.status_code, 'json': r.json()}

    def alert_status(self, alert_id=None):
        _endpoint = self.endpoint("/alerts_feed")
        payload = self.payload
        payload['op'] = 'status'
        payload['alert_id'] = alert_id
        r = requests.get(_endpoint, params=payload)
        return {'status_code': r.status_code, 'json': r.json()}

    def list_alerts(self,):
        _endpoint = self.endpoint("/alerts_feed")
        payload = self.payload
        payload['op'] = 'list'
        r = requests.get(_endpoint, params=payload)
        return {'status_code': r.status_code, 'json': r.json()}

    def read_highlights(self, start_date=None, end_date=None):
        _endpoint = self.endpoint("/highlights_feed")
        payload = self.payload
        if start_date:
            payload['start_date'] = start_date
        if end_date:
            payload['end_date'] = end_date
        r = requests.get(_endpoint, params=payload)
        return {'status_code': r.status_code, 'json': r.json()}

    def post_highlight(self, message, highlight_date=None, source=None):
        _endpoint = self.endpoint("/highlights_post")
        now = datetime.date.today()
        payload = self.payload
        payload['highlight_date'] = highlight_date if highlight_date else now
        payload['description'] = message
        if source:
            payload['source'] = source
        r = requests.post(_endpoint, params=payload)
        return {'status_code': r.status_code, 'json': r.json()}


run.py
import json
import rescuetime

rt = rescuetime.RescueAPI()

print("""
--------------------------------------------------------------------------------
Testing Alerts Feed
--------------------------------------------------------------------------------
""")
result = rt.list_alerts()
print(result['status_code'])
payload = json.dumps(result['json'], indent=2)
print(payload)

print("""
--------------------------------------------------------------------------------
Testing Alerts Status
--------------------------------------------------------------------------------
""")
result = rt.alert_status()
print(result['status_code'])
payload = json.dumps(result['json'], indent=2)
print(payload)

print("""
--------------------------------------------------------------------------------
Testing Highlights Feed
--------------------------------------------------------------------------------
""")
result = rt.read_highlights(end_date="2019-01-23")
print(result['status_code'])
payload = json.dumps(result['json'], indent=2)
print(payload)

print("""
--------------------------------------------------------------------------------
Testing Highlights POST
--------------------------------------------------------------------------------
""")
result = rt.post_highlight("Test")
payload = result['json']
if result['status_code'] == 200:
    print('Success!')
    print(payload)
elif result['status_code'] == 400:
    print('Fail!')
    print(payload)
